### DISCUSSION TITLE
How to apply unique constraint to multiple fields where 1 field is nullable?

### DISCUSSION THREAD

**Discussion by inkeyygnito-dev** [prisma/prisma#21322] [Q&A]
### Question

I am using prisma with postgres and I am trying to apply a unique constraint on multiple fields where 1 field is nullable. I need the nullable field to be compared as well.

For example with my current model:
```prisma
model NetworkAttributionOverride {
  networkAttributionOverrideId Int    @id @default(autoincrement()) @map("network_attribution_override_id")
  networkEntityId              Int    @map("network_entity_id")
  enhancedMetadataId           String @map("enhanced_metadata_id") @db.Uuid
  distributionContextId        Int    @map("distribution_context_id")
  contentEntitlementId         Int    @map("content_entitlement_id")
  playbackTypeId               Int?   @map("playback_type_id")

  entity                                 Entity                                   @relation(fields: [networkEntityId], references: [entityId], onDelete: Cascade)
  enhancedMetadata                       EnhancedMetadata                         @relation(fields: [enhancedMetadataId], references: [enhancedMetadataId], onDelete: Cascade)
  distributionContextRef                 DistributionContextRef                   @relation(fields: [distributionContextId], references: [distributionContextId])
  contentEntitlementRef                  ContentEntitlementRef                    @relation(fields: [contentEntitlementId], references: [contentEntitlementId])
  playbackTypeRef                        PlaybackTypeRef?                         @relation(fields: [playbackTypeId], references: [playbackTypeId])
  networkAttributionOverrideAvailability NetworkAttributionOverrideAvailability[]

  @@unique([enhancedMetadataId, distributionContextId, contentEntitlementId, playbackTypeId])
  @@map("network_attribution_override")
}
```

I'd like these 2 entries to violate the unique constraint.
```
enhancedMetadataId = 1, distributionContextId = 1, contentEntitlementId = 1, playbackTypeId = null
enhancedMetadataId = 1, distributionContextId = 1, contentEntitlementId = 1, playbackTypeId = null
```
But postgres is comparing null != null. In postgres, I can do this by using NULLS NOT DISTINCT:
`CONSTRAINT uq_val1_val2_new UNIQUE NULLS NOT DISTINCT (val1, val2);`

Is there a way I can tell postgres via prisma to ignore comparision of field if both field is null?

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by ludralph** [comment]
Hi @inkeyygnito-dev 👋 

Thank you for raising this question.

Prisma does not currently support the `NULLS NOT DISTINCT` constraint directly in the Prisma schema. You can take a look at the comment in this open issue.

- https://github.com/prisma/prisma/issues/3387#issuecomment-1379686316

However, you can manually add this constraint to your PostgreSQL database and use [`$queryRaw`](https://www.prisma.io/docs/concepts/components/prisma-client/raw-database-access#queryraw) to perform the query. You can also leverage [SafeQL](https://safeql.dev/compatibility/prisma.html) to write raw queries in a type-safe way.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by taylor-shift** [reply]
@ludralph You saved my life today, thank you brother.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
